### DISCUSSION TITLE
x509: certificate signed by unknown authority

### DISCUSSION THREAD

**Discussion by apopaa** [sse-secure-systems/connaisseur#178] [Q&A] [answered]
Hi,

I'm getting the following error when running a signed image built from the connaisseur/setup/Dockerfile (verified by "docker trust inspect"):
`kubectl run -n apns signed2 --image=myregistry.azurecr.io/apopa/testimage2:0.0.1
Error from server (InternalError): Internal error occurred: failed calling webhook "connaisseur-svc.connaisseur.svc": Post https://connaisseur-svc.connaisseur.svc:443/mutate?timeout=30s: x509: certificate signed by unknown authority`

Is the controller trying to make an HTTPS call to the service via the Internet? Because my Kubernetes machines are proxied by a very restrictive corporate proxy, that does not allow arbitrary connections to the Internet.

**Comment by xopham** [comment]
Hi @apopaa, Connaisseur will in general connect via HTTPS. It is hard to judge without knowing the details of your infrastructure, but I might give you a few hints:
First off, you seem to be using ACR from Azure. In general, Connaisseur has been tested with ACR and should work. However, since ACR does not provide a fully standardized notary server, you have to tell Connaisseur that this ACR via the `isACR` flag in `helm/values.yaml`. See here for the specialized guide: https://github.com/sse-secure-systems/connaisseur/tree/master/setup/acr#2-configure-specifics-to-acr
On the other hand, your error message and notes about the proxy seem to indicate that a self-signed certificate may be in use. In this case, you may have to configure the `notary.selfsignedCert` in `helm/values.yaml` as is explained for a local harbor install here: https://github.com/sse-secure-systems/connaisseur/tree/master/setup/acr#2-configure-specifics-to-acr

Let me know if that helps :pray:

**Reply by apopaa** [reply]
I assume that, after I make the changes to selfsignedCert, I have to run a "make upgrade". Right?
I'm asking because it failed:
`bash-5.1# make upgrade
kubectl config set-context --current --namespace connaisseur
Context "kubernetes-admin@kubernetes-epiphany" modified.
helm upgrade connaisseur helm --wait --set alerting.cluster="kubernetes-epiphany"
Error: UPGRADE FAILED: pre-upgrade hooks failed: timed out waiting for the condition
make: *** [Makefile:37: upgrade] Error 1`

**Reply by apopaa** [reply]
Can I just run "make uninstall" followed by "make install" to get to re-install with updated certificates?
Or should I run "make annihilate" instead of "make uninstall"?

**Reply by apopaa** [reply]
Well, I ran "make uninstall" and "make annihilate", and now "make install" fails:
`helm install connaisseur helm --atomic --set alerting.cluster="kubernetes-epiphany"
Error: cannot re-use a name that is still in use
make: *** [Makefile:28: install] Error 1`

Argh, going from bad to worse.

**Reply by xopham** [reply]
I have experienced problems with upgrade in the past as well and would recommend `make uninstall` and `make install` instead. `make annihilate` is meant to help when you run into problems when installing again.

In your current situation, there seems to be some resources remaining. How did you install Connaisseur (via make or helm directly)? Did the `make uninstall` finish successfully? You could check if there is resources in the namespace remaining via:

```bash
kubectl get all -n connaisseur
```

That should list remaining resources. You could also try to delete the namespace as well:

```bash
kubectl delete namespaces connaisseur
```

If that doesn't work, you might need to check if there is any remaining resources.

**Reply by xopham** [reply]
I reconsidered the error message above:

```bash
Internal error occurred: failed calling webhook "connaisseur-svc.connaisseur.svc": Post https://connaisseur-svc.connaisseur.svc:443/mutate?timeout=30s: x509: certificate signed by unknown authority
```

On second thought, it seems unlikely that it is related to the configuration of a selfsigned cert for notary, as the webhook seems to fail calling the Connaisseur service which should be unrelated to any other external network/tls configuration. It seems to indicate that the Connaisseur certificate is not accepted. So either something went wrong when generating the certs for Connaisseur: https://github.com/sse-secure-systems/connaisseur/blob/master/Makefile#L18
Or maybe you directly installed via `helm` and have not prepared the certs via:

```bash
bash helm/certs/gen_certs.sh
```

**Reply by xopham** [reply]
Performed some further testing and noticed that the original error above as also mentioned in my previous response, can result from running `make upgrade`. If that is the case, `make annihilate` does not fully recover form it and if you attempt a new install, you get told `Error: cannot re-use a name that is still in use`. This can be solved by `kubectl delete namespaces connaisseur` after which a re-install should work.

Let me know if that works for you :slightly_smiling_face:

**Reply by xopham** [reply]
tracked under issue #181

**Reply by apopaa** [reply]
I'm using "make", mostly. I've also run helm directly, but only a couple of times and only after running a "make install" before at least once, which I would think generates the certificate (I don't think running gen_certs.sh a second time makes a diff).

As for listing resources in the "connaisseur" namespace and deleting the namespace: I've done both, and there were no resources listed and deleting and recreating the namespace didn't help, "make install" is still failing.

I'm also wondering whether the error message "Error: cannot re-use a name that is still in use" doesn't originate with helm, and not with Kubernetes, ie it is helm caching some object that cannot be recreated with same name, it is not K8s caching some resource.
Unfortunately, running "helm ls -n connaisseur" didn't show any cached helm resources for connaisseur, so my theory doesn't verify.

**Reply by apopaa** [reply]
**At this point I'm trying to clean up my cluster and re-run installation, but am unable to do so.
Here's what I ran:**
bash-5.1# make uninstall
kubectl config set-context --current --namespace connaisseur
Context "kubernetes-admin@kubernetes-epiphany" modified.
helm uninstall connaisseur
Error: warning: Hook pre-delete connaisseur/templates/webhook/role.yaml failed: clusterroles.rbac.authorization.k8s.io "connaisseur-hook-role" already exists
make: *** [Makefile:32: uninstall] Error 1
...
bash-5.1# make annihilate
kubectl delete all,mutatingwebhookconfigurations,clusterroles,clusterrolebindings,configmaps,imagepolicies -lapp.kubernetes.io/instance=connaisseur
pod "connaisseur-bootstrap-sentinel" deleted
pod "connaisseur-deployment-5dc757f498-77h8p" deleted
pod "connaisseur-deployment-5dc757f498-ckjx5" deleted
pod "connaisseur-deployment-5dc757f498-xxzks" deleted
service "connaisseur-svc" deleted
deployment.apps "connaisseur-deployment" deleted
replicaset.apps "connaisseur-deployment-5dc757f498" deleted
clusterrole.rbac.authorization.k8s.io "connaisseur-hook-role" deleted
clusterrole.rbac.authorization.k8s.io "connaisseur-role" deleted
clusterrolebinding.rbac.authorization.k8s.io "connaisseur-hook-rolebinding" deleted
clusterrolebinding.rbac.authorization.k8s.io "connaisseur-rolebinding" deleted
configmap "connaisseur-alert-templates" deleted
configmap "connaisseur-env" deleted
imagepolicy.connaisseur.policy "connaisseur-policy" deleted

**This left the "connaisseur" namespace behind, so on the K8s master I ran:**
[user@masternode ~]$ kubectl delete ns connaisseur
namespace "connaisseur" deleted

**I checked that the "connaisseur" namespace was gone, and that no clusterroles with "connaisseur" in their names existed, then I re-ran:**
bash-5.1# make install
bash helm/certs/gen_certs.sh
Generating RSA private key, 4096 bit long modulus (2 primes)
................................................................................................................................................................................................................++++
.......................................++++
e is 65537 (0x010001)
Signature ok
subject=CN = connaisseur-svc.connaisseur.svc
Getting Private key
kubectl create ns connaisseur || true
namespace/connaisseur created
kubectl config set-context --current --namespace connaisseur
Context "kubernetes-admin@kubernetes-epiphany" modified.
.
.=============================================
.
. The installation may last up to 5 minutes.
.
.=============================================
.
helm install connaisseur helm --atomic --set alerting.cluster="kubernetes-epiphany"
Error: an error occurred while uninstalling the release. original install error: timed out waiting for the condition: timed out waiting for the condition
make: *** [Makefile:28: install] Error 1

**Reply by apopaa** [reply]
I might have found the problem. Since I can't fix the current installation, I thought I annihilate it and delete the "connaisseur" K8s namespace, and do a new installation in a brand new K8s namespace named "connaisseur1". I ran make install and got this error:

bash-5.1# make install NAMESPACE=connaisseur1
bash helm/certs/gen_certs.sh
Generating RSA private key, 4096 bit long modulus (2 primes)
...
subject=CN = connaisseur-svc.connaisseur.svc
Getting Private key
kubectl create ns connaisseur1 || true
namespace/connaisseur1 created
kubectl config set-context --current --namespace connaisseur1
Context "kubernetes-admin@kubernetes-epiphany" modified.
...
helm install connaisseur helm --atomic --set alerting.cluster="kubernetes-epiphany"
Error: rendered manifests contain a resource that already exists. Unable to continue with install: Namespace "apns" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-namespace" must equal "connaisseur1": current value is "connaisseur"
make: *** [Makefile:28: install] Error 1

"apns" is the K8s namespace where I want to deploy my (un)signed test images, and that I have set inside helm/values.yml as the targetNamespaces:
targetNamespaces: ['apns']

**Reply by xopham** [reply]
I see what you mean. "targetNamespaces" was introduced for namespaced validation and would restrict validation to namespaces specified therein. To do so, it would create the namespaces when doing the install and if the namespace remains it could block a new installation since it already exists.
We have changed the namespaced validation for the next release to improve this behavior: https://sse-secure-systems.github.io/connaisseur/develop/features/namespaced_validation/

**Reply by apopaa** [reply]
Damn, "targetNamespaces" was such a convenient quick way of verifying ALL my images, since they all get deployed to the same namespace. I have to use the "policy" feature and provide a pattern to match my images.

Still being unable to install, though. Getting this error now:
bash-5.1# make install
...
helm install connaisseur helm --atomic --set alerting.cluster="kubernetes-epiphany"
Error: an error occurred while uninstalling the release. original install error: timed out waiting for the condition: timed out waiting for the condition
make: *** [Makefile:28: install] Error 1

What condition is it waiting for? Any error log somewhere that could tell me what's the "condition"?

**Reply by apopaa** [reply]
OK, the "condition" it is waiting for is the readiness check of the connaisseur pods: it is failing repeatedly with error 500; the liveness check otoh succeeds. 

This is the end of the log of one of the pods; no errors:
[2021-07-22 05:01:34,296] INFO: 10.244.3.1 - - [22/Jul/2021 05:01:34] "GET /health HTTP/1.1" 200 -
[2021-07-22 05:01:39,175] INFO: 10.244.3.1 - - [22/Jul/2021 05:01:39] "GET /ready HTTP/1.1" 500 -
[2021-07-22 05:01:44,810] INFO: 10.244.3.1 - - [22/Jul/2021 05:01:44] "GET /health HTTP/1.1" 200 -
[2021-07-22 05:01:44,813] INFO: 10.244.3.1 - - [22/Jul/2021 05:01:44] "GET /ready HTTP/1.1" 500 -
[2021-07-22 05:01:48,001] INFO: 10.244.3.1 - - [22/Jul/2021 05:01:48] "GET /ready HTTP/1.1" 500 -
[2021-07-22 05:01:49,571] INFO: 10.244.3.1 - - [22/Jul/2021 05:01:49] "GET /health HTTP/1.1" 200 -

**Reply by xopham** [reply]
Don't get me wrong, namespaced validation will continue to work, just better. It is not removed, but the way it is implemented rn on each install the namespaces to limit validation have to be created on each install and if they already exist...install will fail. In the 2.0 release, you will be able to activate it and then use labels to mark namespaces for validation.

To your errors: It is a bit hard to judge from the healthiness and readiness alone. However, since you use ACR have you set `notary.isAcr: true` in `helm/values.yaml`? ACR does not provide a proper notary health endpoint and therefore Connaisseur has to ignore notary health in that case which is achieved by the `isAcr` key. See here: https://github.com/sse-secure-systems/connaisseur/blob/master/helm/values.yaml#L54

**Reply by apopaa** [reply]
Yes, isAcr is true:

  \# if you use Azure Container Registry (ACR) for your notary
  \# changes some behaviour, such as health probes and how to retrieve auth tokens
  \# for compatibility with ACR set to `true`
  isAcr: true

Besides, in my log the health probe returns 200; it is the ready probe that returns 500.

Update: It appears like the problems with the ready probe were of a transitory nature due (probably) to my network infrastructure being slow yesterday and timing out certain deployments, resulting in the Connaisseur controller not properly setting up. Today the network is fast again and Connaisseur deployed without problems.

**Reply by apopaa** [reply]
OK, so now that the deployment problems have been figured out, we're back to the opening issue: "x509: certificate signed by unknown authority"

My attempt to deploy an unsigned (testimage2) and a signed (testimage3) image:
_[user@master ~]$ kubectl run -n apns unsigned2 --image=myregistry.azurecr.io/apopa/testimage2:0.0.1
Error from server (InternalError): Internal error occurred: failed calling webhook "connaisseur-svc.connaisseur.svc": Post https://connaisseur-svc.connaisseur.svc:443/mutate?timeout=30s: x509: certificate signed by unknown authority
[user@master ~]$ kubectl run -n apns signed3 --image=myregistry.azurecr.io/apopa/testimage3:1.0.1
Error from server (InternalError): Internal error occurred: failed calling webhook "connaisseur-svc.connaisseur.svc": Post https://connaisseur-svc.connaisseur.svc:443/mutate?timeout=30s: x509: certificate signed by unknown authority_

**Comment by apopaa** [comment, marked as answer]
OK, so I've made some progress in that, as I had suspected, the "x509: certificate signed by unknown authority" error was related to my proxy, which is configured to inspect content and so decrypts any TLS traffic and then re-encrypts it with its own certificate, which of course is then seen as "unknown" by the connaisseur webhook. The solution was pretty simple: add ".svc" to the no_proxy environment variable in /etc/kubernetes/manifests/kube-apiserver.yaml, to make kube-apiserver bypass the proxy.

Now, in doing that, I think I uncovered a bug in the handling of the policy configuration from values.yml: because I changed the manifest of kube-apiserver, the kubelet attempted to reload it, but failed with the following error in systemd logs (journalctl):

Jul 22 16:17:56 masternode kubelet[38796]: E0722 16:17:56.313082 38796 kubelet.go:1664] Failed creating a mirror pod for "kube-apiserver-masternode_kube-system(6da95f69ab0a006e91f81c7c5d98f14f)": admission webhook "connaisseur-svc.connaisseur.svc" denied the request: "masternode:5000/k8s.gcr.io/kube-apiserver:v1.18.6" is not a valid image format.

That's because my K8s is set up to load system pods from a local repo installed on masternode machine, and no pattern in the policy in values.yml was matching "masternode:5000/k8s.gcr.io/kube-apiserver:v1.18.6", and so connaisseur rejected loading the image. I had expected it to accept any image it cannot match, by default; turns out it is rejects any image it cannot match, by default. Is this a bug, or a feature?
The workaround is simple: add a default policy for ":" with "verify: false", in addition to the specific policy that validates my images:
policy:

pattern: "myregistry.azurecr.io/apopa/:"
verify: true
pattern: ":"
verify: false

**Reply by xopham** [reply]
Hey @apopaa, happy to hear things are working now :slightly_smiling_face: 
To your question about unmatched images: We had that discussion and decided for the secure default and wanted to enforce explicit exclusion as you did. Therefore, we would not consider it a bug.

**Reply by apopaa** [reply]
Actually, I think you have a bug in your 1.5 code because even with this setting in values.yml:
_policy:
  \- pattern: "*"
    verify: false_
it still denies the request:
_denied the request: "master-node:5000/k8s.gcr.io/kube-apiserver:v1.18.6" is not a valid image format._

I think the double ":" throws your code off.

I see 2.0 is out today, so this bug might be unimportant now.

**Reply by xopham** [reply]
I think you might be correct and we have a bug there in the image format validation that still persists into version 2.0. Connaisseur seems to reject the two '.' in the "repository" part. Escalating to issue now :arrow_up:

**Reply by xopham** [reply]
Issue #194 . We'll be looking into this.

**Reply by phbelitz** [reply]
> Actually, I think you have a bug in your 1.5 code because even with this setting in values.yml:
> _policy: - pattern: "*" verify: false_
> it still denies the request:
> _denied the request: "master-node:5000/k8s.gcr.io/kube-apiserver:v1.18.6" is not a valid image format._
> 
> I think the double ":" throws your code off.
> 
> I see 2.0 is out today, so this bug might be unimportant now.

Hey there! I assume you signed your image using:

```bash
DOCKER_CONTENT_TRUST=1 docker image push master-node:5000/k8s.gcr.io/kube-apiserver:v1.18.6
```

Under which of these two URL's is your trust data available?

```bash
https://<your-notary-url>/v2/master-node:5000/k8s.gcr.io/kube-apiserver/_trust/tuf/root.json
# OR
https://<your-notary-url>/v2/k8s.gcr.io/kube-apiserver/_trust/tuf/root.json
```

**Reply by apopaa** [reply]
That image is not signed. And I want to keep it that way, at least for now. All I want atm is to verify the images containing my application, ie "myregistry/apopa/testimage2" and the like.

**Reply by phbelitz** [reply]
Well well well! We fixed the issue with our new v2.1.0 release. Now your ` master-node:5000/k8s.gcr.io/kube-apiserver:v1.18.6` image should be correctly accepted as a valid image.

CHEERS! :wine_glass:
